src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1759,6 +1762,415 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
     }
 }
 
+UnorderedSet<LedgerKey>
+LedgerManagerImpl::getReadWriteKeysForStage(ApplyStage const& stage)
+{
+    UnorderedSet<LedgerKey> res;
+
+    for (auto const& txs : stage)
```

Comment:
terminology is getting tricky! phase, stage, bundle are all starting to sound a bit similar. I wonder if there’s a way to abstract it away via an iterator of sorts if all you care about is iterating over all transactions, for example. We can pass a lambda to the iterator. This way we can try to minimize places in core that need to be aware of the internal structure of a parallel tx set.

Reply:
just for context, I noticed there are lots of nested loops in the code making it harder to review, so I wonder if we can create an abstraction for tx sets to avoid bleeding internal details into other subsystems of core.

Reply:
I added an iterator to ApplyStage. This gets rid of all but one of the nested loops. I didn't add something similar for the entire phase, but I don't think that's necessary here. Let me know what you think.